tracing/plugin/otlp.go

```diff
@@ -0,0 +1,143 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package plugin
+
+import (
+	"io"
+
+	"github.com/containerd/containerd/log"
+	"github.com/containerd/containerd/plugin"
+	"github.com/pkg/errors"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
+	"go.opentelemetry.io/otel/propagation"
+	"go.opentelemetry.io/otel/sdk/resource"
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+	semconv "go.opentelemetry.io/otel/semconv/v1.4.0"
+	"google.golang.org/grpc"
+)
+
+const exporterPlugin = "otlp"
+
+func init() {
+	plugin.Register(&plugin.Registration{
+		ID:     exporterPlugin,
+		Type:   plugin.TracingProcessorPlugin,
+		Config: &OTLPConfig{},
+		InitFn: func(ic *plugin.InitContext) (interface{}, error) {
+			cfg := ic.Config.(*OTLPConfig)
+			if cfg.ExporterEndpoint == "" {
+				return nil, errors.Wrap(plugin.ErrSkipPlugin, "otlp endpoint not set")
+			}
+			dialOpts := []grpc.DialOption{grpc.WithBlock()}
+			if cfg.Insecure {
+				dialOpts = append(dialOpts, grpc.WithInsecure())
+			}
+
+			exp, err := otlptracegrpc.New(ic.Context,
+				otlptracegrpc.WithEndpoint(cfg.ExporterEndpoint),
+				otlptracegrpc.WithDialOption(dialOpts...),
+			)
+			if err != nil {
+				return nil, errors.Wrap(err, "failed to create otlp exporter")
+			}
+			return sdktrace.NewBatchSpanProcessor(exp), nil
+		},
+	})
+	plugin.Register(&plugin.Registration{
+		ID:       "tracing",
+		Type:     plugin.InternalPlugin,
+		Requires: []plugin.Type{plugin.TracingProcessorPlugin},
+		Config:   &TraceConfig{ServiceName: "containerd"},
+		InitFn: func(ic *plugin.InitContext) (interface{}, error) {
+			return newTracer(ic)
+		},
+	})
+}
+
+// OTLPConfig holds the configurations for the built-in otlp span processor
+type OTLPConfig struct {
+	ExporterEndpoint string `toml:"endpoint"`
```

Comment:
Could the variable name be `Endpoint` as well? Would it have endpoints other than the exporter endpoint?

Reply:
Fixed.